# 🚀 E-COMMERCE PROFITABILITY & FULFILLMENT PROJECT

> 📊 Turning E-Commerce Data into Actionable Business Insights

---

## 👨‍💻 AUTHOR

**Prakrit Kumar**

**Role:** Data Analyst  
**Data Cleaning:** Python (Pandas, NumPy)  
**Start Date:** 2 | September | 2026

---

In [1]:
import pandas as pd
import numpy as np

# for data clening 

In [2]:
data=pd.ExcelFile("Raw_data.xlsx")
print(data.sheet_names)

# first see all tables name

['FactOrderLine', 'DimDate', 'DimProduct', 'DimCustomer', 'DimGeography', 'DimPromotion', 'DimFulfillment', 'DimSalesChannel', 'DimReturnReason', 'DimCohortAge', 'Data Dictionary']


In [3]:
sheets = pd.read_excel("Raw_data.xlsx", sheet_name=None)

# read aal excel file

FactOrderLine_df = sheets["FactOrderLine"]
DimDate_df = sheets["DimDate"]
DimProduct_df = sheets["DimProduct"]
DimCustomer_df = sheets["DimCustomer"]
DimGeography_df = sheets["DimGeography"]
DimPromotion_df = sheets["DimPromotion"]
DimFulfillment_df = sheets["DimFulfillment"]
DimSalesChannel_df = sheets["DimSalesChannel"]
DimReturnReason_df = sheets["DimReturnReason"]
DimCohortAge_df = sheets["DimCohortAge"]
DataDictionary_df = sheets["Data Dictionary"]

# **clening process start**

### **1. FactOrderLine_df clening start** 

#### Cheak head

In [4]:
FactOrderLine_df.head()

,LineID,OrderID,OrderDateKey,ShipDateKey,DeliveryDateKey,ReturnDateKey,ProductKey,CustomerKey,GeographyKey,PromotionKey,...,ReturnProcessingCost,ContributionRevenue,ContributionMargin,LostSalesValue,ReturnLoss,PromisedDeliveryDays,ActualDeliveryDays,OnTimeFlag,IsPromoted,Currency
0,L0000001,O000001,20250329,20250329,20250401.0,NaN,P043,C00001,G021,PR08,...,0.0,35.8105,18.2005,0.0,0.0,3,3.0,1.0,1,EUR
1,L0000002,O000001,20250329,20250329,20250401.0,NaN,P055,C00001,G021,PR00,...,0.0,160.8800,65.1000,0.0,0.0,3,3.0,1.0,0,EUR
2,L0000003,O000002,20250714,20250715,20250717.0,NaN,P031,C00001,G021,PR00,...,0.0,69.4600,27.7200,0.0,0.0,3,2.0,1.0,0,EUR
3,L0000004,O000003,20250926,20250927,20250930.0,NaN,P053,C00001,G021,PR00,...,0.0,52.0300,20.3500,0.0,0.0,3,3.0,1.0,0,EUR
4,L0000005,O000004,20250315,20250315,20250319.0,NaN,P006,C00002,G009,PR08,...,0.0,106.4460,3.3260,0.0,0.0,4,4.0,1.0,1,EUR


#### cheak info

In [5]:
FactOrderLine_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9109 entries, 0 to 9108
Data columns (total 46 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   LineID                  9109 non-null   object 
 1   OrderID                 9109 non-null   object 
 2   OrderDateKey            9109 non-null   int64  
 3   ShipDateKey             9109 non-null   int64  
 4   DeliveryDateKey         8967 non-null   float64
 5   ReturnDateKey           1046 non-null   float64
 6   ProductKey              9109 non-null   object 
 7   CustomerKey             9109 non-null   object 
 8   GeographyKey            9109 non-null   object 
 9   PromotionKey            9109 non-null   object 
 10  FulfillmentKey          9109 non-null   object 
 11  SalesChannelKey         9109 non-null   object 
 12  ReturnReasonKey         9109 non-null   object 
 13  CohortAgeKey            9109 non-null   int64  
 14  OrderStatus             9109 non-null   

#### chek null values

In [6]:
FactOrderLine_df.isnull().sum()

LineID                       0
OrderID                      0
OrderDateKey                 0
ShipDateKey                  0
DeliveryDateKey            142
ReturnDateKey             8063
ProductKey                   0
CustomerKey                  0
GeographyKey                 0
PromotionKey                 0
FulfillmentKey               0
SalesChannelKey              0
ReturnReasonKey              0
CohortAgeKey                 0
OrderStatus                  0
OrderedQuantity              0
FulfilledQuantity            0
ReturnedQuantity             0
OpeningStockUnits            0
StockStatus                  0
UnitListPrice                0
UnitCost                     0
AppliedDiscountPct           0
ReturnRecoveryPct            0
GrossSales                   0
DiscountAmount               0
SalesBeforeReturns           0
RefundAmount                 0
NetSales                     0
ShippingRevenue              0
ProductCost                  0
OutboundShippingCost         0
ReturnSh

#### chek statistics

In [7]:
FactOrderLine_df.describe()

,OrderDateKey,ShipDateKey,DeliveryDateKey,ReturnDateKey,CohortAgeKey,OrderedQuantity,FulfilledQuantity,ReturnedQuantity,OpeningStockUnits,UnitListPrice,...,AllocatedMarketingCost,ReturnProcessingCost,ContributionRevenue,ContributionMargin,LostSalesValue,ReturnLoss,PromisedDeliveryDays,ActualDeliveryDays,OnTimeFlag,IsPromoted
count,9.109000e+03,9.109000e+03,8.967000e+03,1.046000e+03,9109.000000,9109.000000,9109.000000,9109.000000,9109.000000,9109.000000,...,9109.000000,9109.000000,9109.000000,9109.000000,9109.000000,9109.000000,9109.000000,8967.000000,8967.000000,9109.000000
mean,2.024702e+07,2.024704e+07,2.024713e+07,2.024736e+07,3.542431,1.256889,1.202218,0.120650,18.428916,74.366337,...,3.280315,0.400401,72.149184,14.013821,4.030626,11.296341,3.557580,3.319728,0.927512,0.679548
std,4.772186e+03,4.793408e+03,4.849853e+03,5.100766e+03,3.674277,0.503706,0.554012,0.345052,10.665108,33.667550,...,5.158309,1.137040,52.325539,30.123466,20.289225,36.881618,0.965765,1.508040,0.259309,0.466676
min,2.024010e+07,2.024010e+07,2.024010e+07,2.024011e+07,0.000000,1.000000,0.000000,0.000000,0.000000,20.520000,...,0.000000,0.000000,0.000000,-270.640000,0.000000,0.000000,2.000000,1.000000,0.000000,0.000000
25%,2.024101e+07,2.024101e+07,2.024102e+07,2.024102e+07,0.000000,1.000000,1.000000,0.000000,9.000000,47.630000,...,0.490000,0.000000,41.026400,1.208000,0.000000,0.000000,3.000000,2.000000,1.000000,0.000000
50%,2.025032e+07,2.025032e+07,2.025032e+07,2.025040e+07,3.000000,1.000000,1.000000,0.000000,18.000000,68.040000,...,1.070000,0.000000,64.857500,14.628800,0.000000,0.000000,3.000000,3.000000,1.000000,1.000000
75%,2.025073e+07,2.025073e+07,2.025080e+07,2.025082e+07,6.000000,1.000000,1.000000,0.000000,28.000000,95.200000,...,3.000000,0.000000,97.140000,28.300000,0.000000,0.000000,4.000000,4.000000,1.000000,1.000000
max,2.025123e+07,2.026010e+07,2.026010e+07,2.026013e+07,17.000000,3.000000,3.000000,3.000000,38.000000,199.300000,...,23.000000,6.850000,446.430000,168.344400,309.880000,546.940000,6.000000,11.000000,1.000000,1.000000


#### cheak duplicate

In [8]:
FactOrderLine_df.duplicated().sum()

np.int64(0)

#### cheak shape of the data

In [9]:
print("total row =",FactOrderLine_df.shape[0])
print("total columns =",FactOrderLine_df.shape[1])

total row = 9109
total columns = 46


### **About the Dataframe**
##### **Null Values**
- DeliveryDateKey = 142
- ReturnDateKey = 8063
- ActualDeliveryDays = 142
- OnTimeFlag = 142
##### No duplicate Values
##### Total Row = 9109
##### Covert Column Name Into In Lower Case
#### **Not Needed to Fill Null Values Because They Are Meaningful Nulls**
- DeliveryDateKey`, `ActualDeliveryDays`, `OnTimeFlag` (142 nulls) — order was cancelled due to stockout before delivery,
  so no delivery occurred.
- `ReturnDateKey` (8063 nulls) — order was never returned, so no return date exists.

In [10]:
# Cheak the Cancelled - Stockout status
FactOrderLine_df[FactOrderLine_df['DeliveryDateKey'].isnull()]['OrderStatus'].value_counts()

OrderStatus
Cancelled - stockout    142
Name: count, dtype: int64

In [11]:
# Cheak Column Name case
FactOrderLine_df.columns

Index(['LineID', 'OrderID', 'OrderDateKey', 'ShipDateKey', 'DeliveryDateKey',
       'ReturnDateKey', 'ProductKey', 'CustomerKey', 'GeographyKey',
       'PromotionKey', 'FulfillmentKey', 'SalesChannelKey', 'ReturnReasonKey',
       'CohortAgeKey', 'OrderStatus', 'OrderedQuantity', 'FulfilledQuantity',
       'ReturnedQuantity', 'OpeningStockUnits', 'StockStatus', 'UnitListPrice',
       'UnitCost', 'AppliedDiscountPct', 'ReturnRecoveryPct', 'GrossSales',
       'DiscountAmount', 'SalesBeforeReturns', 'RefundAmount', 'NetSales',
       'ShippingRevenue', 'ProductCost', 'OutboundShippingCost',
       'ReturnShippingCost', 'FulfillmentCost', 'PaymentFee',
       'AllocatedMarketingCost', 'ReturnProcessingCost', 'ContributionRevenue',
       'ContributionMargin', 'LostSalesValue', 'ReturnLoss',
       'PromisedDeliveryDays', 'ActualDeliveryDays', 'OnTimeFlag',
       'IsPromoted', 'Currency'],
      dtype='object')

In [12]:
# Conver into Lower Case
FactOrderLine_df.columns = FactOrderLine_df.columns.str.lower()

In [13]:
# Cheak Again
FactOrderLine_df.columns

Index(['lineid', 'orderid', 'orderdatekey', 'shipdatekey', 'deliverydatekey',
       'returndatekey', 'productkey', 'customerkey', 'geographykey',
       'promotionkey', 'fulfillmentkey', 'saleschannelkey', 'returnreasonkey',
       'cohortagekey', 'orderstatus', 'orderedquantity', 'fulfilledquantity',
       'returnedquantity', 'openingstockunits', 'stockstatus', 'unitlistprice',
       'unitcost', 'applieddiscountpct', 'returnrecoverypct', 'grosssales',
       'discountamount', 'salesbeforereturns', 'refundamount', 'netsales',
       'shippingrevenue', 'productcost', 'outboundshippingcost',
       'returnshippingcost', 'fulfillmentcost', 'paymentfee',
       'allocatedmarketingcost', 'returnprocessingcost', 'contributionrevenue',
       'contributionmargin', 'lostsalesvalue', 'returnloss',
       'promiseddeliverydays', 'actualdeliverydays', 'ontimeflag',
       'ispromoted', 'currency'],
      dtype='object')

### FactOrderLine_df is Done...!!

### **2. DimDate_df clening start**

#### cheak head

In [14]:
DimDate_df

,DateKey,Date,Year,Quarter,YearQuarter,MonthNumber,Month,YearMonth,WeekNumber,YearWeek,Day,DayOfWeekNumber,DayOfWeek,DayType
0,20240101,2024-01-01,2024,Q1,2024-Q1,1,January,2024-01,1,2024-W01,1,1,Monday,Weekday
1,20240102,2024-01-02,2024,Q1,2024-Q1,1,January,2024-01,1,2024-W01,2,2,Tuesday,Weekday
2,20240103,2024-01-03,2024,Q1,2024-Q1,1,January,2024-01,1,2024-W01,3,3,Wednesday,Weekday
3,20240104,2024-01-04,2024,Q1,2024-Q1,1,January,2024-01,1,2024-W01,4,4,Thursday,Weekday
4,20240105,2024-01-05,2024,Q1,2024-Q1,1,January,2024-01,1,2024-W01,5,5,Friday,Weekday
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1091,20261227,2026-12-27,2026,Q4,2026-Q4,12,December,2026-12,52,2026-W52,27,7,Sunday,Weekend
1092,20261228,2026-12-28,2026,Q4,2026-Q4,12,December,2026-12,53,2026-W53,28,1,Monday,Weekday
1093,20261229,2026-12-29,2026,Q4,2026-Q4,12,December,2026-12,53,2026-W53,29,2,Tuesday,Weekday
1094,20261230,2026-12-30,2026,Q4,2026-Q4,12,December,2026-12,53,2026-W53,30,3,Wednesday,Weekday


#### cheak info

In [15]:
DimDate_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1096 entries, 0 to 1095
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   DateKey          1096 non-null   int64         
 1   Date             1096 non-null   datetime64[ns]
 2   Year             1096 non-null   int64         
 3   Quarter          1096 non-null   object        
 4   YearQuarter      1096 non-null   object        
 5   MonthNumber      1096 non-null   int64         
 6   Month            1096 non-null   object        
 7   YearMonth        1096 non-null   object        
 8   WeekNumber       1096 non-null   int64         
 9   YearWeek         1096 non-null   object        
 10  Day              1096 non-null   int64         
 11  DayOfWeekNumber  1096 non-null   int64         
 12  DayOfWeek        1096 non-null   object        
 13  DayType          1096 non-null   object        
dtypes: datetime64[ns](1), int64(6), object(7

#### cheak null values

In [16]:
DimDate_df.isnull().sum()

DateKey            0
Date               0
Year               0
Quarter            0
YearQuarter        0
MonthNumber        0
Month              0
YearMonth          0
WeekNumber         0
YearWeek           0
Day                0
DayOfWeekNumber    0
DayOfWeek          0
DayType            0
dtype: int64

#### cheak statistics

In [17]:
DimDate_df.describe()

,DateKey,Date,Year,MonthNumber,WeekNumber,Day,DayOfWeekNumber
count,1.096000e+03,1096,1096.000000,1096.000000,1096.000000,1096.000000,1096.000000
mean,2.025066e+07,2025-07-01 12:00:00,2024.999088,6.521898,26.833942,15.732664,3.994526
min,2.024010e+07,2024-01-01 00:00:00,2024.000000,1.000000,1.000000,1.000000,1.000000
25%,2.024098e+07,2024-09-30 18:00:00,2024.000000,4.000000,14.000000,8.000000,2.000000
50%,2.025070e+07,2025-07-01 12:00:00,2025.000000,7.000000,27.000000,16.000000,4.000000
75%,2.026040e+07,2026-04-01 06:00:00,2026.000000,10.000000,40.000000,23.000000,6.000000
max,2.026123e+07,2026-12-31 00:00:00,2026.000000,12.000000,53.000000,31.000000,7.000000
std,8.178333e+03,NaN,0.817055,3.450561,15.077268,8.805388,2.000449


#### cheak duplicate

In [18]:
DimDate_df.duplicated().sum()

np.int64(0)

#### cheak shape of the data

In [19]:
print("total row =",DimDate_df.shape[0])
print("total columns =",DimDate_df.shape[1])

total row = 1096
total columns = 14


#### **About the Dataframe**
- No Null values
- No Duplicate values
- Dta Type Is Correct
- Total Row is = 1096
- Total Column =14
- Conver Column Name In Lower Case

In [20]:
# Cheak Column Case
DimDate_df.columns

Index(['DateKey', 'Date', 'Year', 'Quarter', 'YearQuarter', 'MonthNumber',
       'Month', 'YearMonth', 'WeekNumber', 'YearWeek', 'Day',
       'DayOfWeekNumber', 'DayOfWeek', 'DayType'],
      dtype='object')

In [21]:
# Convert into Lower Case
DimDate_df.columns = DimDate_df.columns.str.lower()

In [22]:
# Cheak again
DimDate_df.columns

Index(['datekey', 'date', 'year', 'quarter', 'yearquarter', 'monthnumber',
       'month', 'yearmonth', 'weeknumber', 'yearweek', 'day',
       'dayofweeknumber', 'dayofweek', 'daytype'],
      dtype='object')

### DimDate Clening Is Done..!!

### **3. DimProduct_df clening start**

#### cheak head

In [23]:
DimProduct_df.head()

,ProductKey,SKU,Department,Category,Subcategory,ProductName,Brand,BrandTier,ListPrice,StandardUnitCost,WeightClass,ReturnRisk,ReorderLeadDays,Seasonality
0,P001,SKU-ELE-001,Personal & Lifestyle,Electronics,Audio,Tech Audio Essential,Tech,Value,65.24,45.48,Medium,0.08,35,Q4
1,P002,SKU-ELE-002,Personal & Lifestyle,Electronics,Audio,Tech Audio Plus,Tech,Core,95.39,67.60,Medium,0.08,35,Q4
2,P003,SKU-ELE-003,Personal & Lifestyle,Electronics,Audio,Tech Audio Pro,Tech Premium,Core,128.01,89.73,Medium,0.08,35,Q4
3,P004,SKU-ELE-004,Personal & Lifestyle,Electronics,Audio,Tech Audio Elite,Tech Premium,Premium,157.26,108.22,Medium,0.08,35,Q4
4,P005,SKU-ELE-005,Personal & Lifestyle,Electronics,Wearables,Tech Wearables Essential,Tech,Value,66.52,43.79,Medium,0.08,35,Q4


#### cheak info

In [24]:
DimProduct_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60 entries, 0 to 59
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ProductKey        60 non-null     object 
 1   SKU               60 non-null     object 
 2   Department        60 non-null     object 
 3   Category          60 non-null     object 
 4   Subcategory       60 non-null     object 
 5   ProductName       60 non-null     object 
 6   Brand             60 non-null     object 
 7   BrandTier         60 non-null     object 
 8   ListPrice         60 non-null     float64
 9   StandardUnitCost  60 non-null     float64
 10  WeightClass       60 non-null     object 
 11  ReturnRisk        60 non-null     float64
 12  ReorderLeadDays   60 non-null     int64  
 13  Seasonality       60 non-null     object 
dtypes: float64(3), int64(1), object(10)
memory usage: 6.7+ KB


#### cheak null values

In [25]:
DimProduct_df.isnull().sum()

ProductKey          0
SKU                 0
Department          0
Category            0
Subcategory         0
ProductName         0
Brand               0
BrandTier           0
ListPrice           0
StandardUnitCost    0
WeightClass         0
ReturnRisk          0
ReorderLeadDays     0
Seasonality         0
dtype: int64

#### cheak statistics

In [26]:
DimProduct_df.describe()

,ListPrice,StandardUnitCost,ReturnRisk,ReorderLeadDays
count,60.000000,60.000000,60.000000,60.000000
mean,75.766500,38.173500,0.103000,26.000000
std,34.718225,26.822167,0.071564,6.016925
min,26.350000,7.150000,0.035000,18.000000
25%,48.502500,19.112500,0.070000,21.000000
50%,69.925000,31.755000,0.080000,28.000000
75%,94.955000,43.990000,0.090000,28.000000
max,173.230000,122.750000,0.240000,35.000000


#### cheak duplicate

In [27]:
DimProduct_df.duplicated().sum()

np.int64(0)

#### cheak shape of the data

In [28]:
print("total row =",DimProduct_df.shape[0])
print("total columns =",DimProduct_df.shape[1])

total row = 60
total columns = 14


#### **About the Dataframe**
- No Null values
- No Duplicate values
- Dta Type Is Correct
- Total Row is = 60
- Total Column = 14
- Convert Column Name in Lower Case


In [29]:
# Cheak Column Name Case
DimProduct_df.columns

Index(['ProductKey', 'SKU', 'Department', 'Category', 'Subcategory',
       'ProductName', 'Brand', 'BrandTier', 'ListPrice', 'StandardUnitCost',
       'WeightClass', 'ReturnRisk', 'ReorderLeadDays', 'Seasonality'],
      dtype='object')

In [103]:
# convert Column Name in Lower Case
DimProduct_df.columns = DimProduct_df.columns.str.lower()

In [104]:
# Cheak Again
DimProduct_df.columns

Index(['productkey', 'sku', 'department', 'category', 'subcategory',
       'productname', 'brand', 'brandtier', 'listprice', 'standardunitcost',
       'weightclass', 'returnrisk', 'reorderleaddays', 'seasonality'],
      dtype='object')

### DimProduct_df Clening Is Done..!!

### **4. DimCustomer_df clening start**

#### cheak head

In [32]:
DimCustomer_df.head()

,CustomerKey,CustomerLabel,CustomerSegment,LoyaltyTier,AcquisitionChannel,FirstPurchaseDateKey,CohortMonth,CohortYear,CohortQuarter,HomeGeographyKey,PreferredCategory,AgeBand
0,C00001,Customer 00001,Growth,Silver,Email / CRM,20250329,2025-03,2025,2025-Q1,G021,Sports & Outdoors,25–34
1,C00002,Customer 00002,At Risk,NaN,Paid Search,20250315,2025-03,2025,2025-Q1,G009,Electronics,35–44
2,C00003,Customer 00003,Deal Seeker,NaN,Paid Search,20240528,2024-05,2024,2024-Q2,G005,Fashion,25–34
3,C00004,Customer 00004,New,NaN,Email / CRM,20240813,2024-08,2024,2024-Q3,G023,Electronics,25–34
4,C00005,Customer 00005,New,NaN,Paid Search,20241122,2024-11,2024,2024-Q4,G006,Electronics,35–44


#### cheak info

In [33]:
DimCustomer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2200 entries, 0 to 2199
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   CustomerKey           2200 non-null   object
 1   CustomerLabel         2200 non-null   object
 2   CustomerSegment       2200 non-null   object
 3   LoyaltyTier           1034 non-null   object
 4   AcquisitionChannel    2200 non-null   object
 5   FirstPurchaseDateKey  2200 non-null   int64 
 6   CohortMonth           2200 non-null   object
 7   CohortYear            2200 non-null   int64 
 8   CohortQuarter         2200 non-null   object
 9   HomeGeographyKey      2200 non-null   object
 10  PreferredCategory     2200 non-null   object
 11  AgeBand               2200 non-null   object
dtypes: int64(2), object(10)
memory usage: 206.4+ KB


#### cheak null values

In [34]:
DimCustomer_df.isnull().sum()

CustomerKey                0
CustomerLabel              0
CustomerSegment            0
LoyaltyTier             1166
AcquisitionChannel         0
FirstPurchaseDateKey       0
CohortMonth                0
CohortYear                 0
CohortQuarter              0
HomeGeographyKey           0
PreferredCategory          0
AgeBand                    0
dtype: int64

#### cheak statistics

In [35]:
DimCustomer_df.describe()

,FirstPurchaseDateKey,CohortYear
count,2.200000e+03,2200.000000
mean,2.024549e+07,2024.486364
std,4.946085e+03,0.499928
min,2.024010e+07,2024.000000
25%,2.024063e+07,2024.000000
50%,2.024123e+07,2024.000000
75%,2.025052e+07,2025.000000
max,2.025103e+07,2025.000000


#### cheak duplicate

In [36]:
DimCustomer_df.duplicated().sum()

np.int64(0)

#### cheak shape of the data

In [37]:
print("total row =",DimCustomer_df.shape[0])
print("total columns =",DimCustomer_df.shape[1])

total row = 2200
total columns = 12


#### **About the Dataframe**
##### Null Values
- LoyalaltyTier = 1166
##### DataType is Correct
##### No Duplicate values
##### total row = 2200
##### Total column = 12
##### Convert column name in lower case

In [38]:
# Fill Null Value In LoyaltyTier to ("Not_Available")
DimCustomer_df['LoyaltyTier'] = DimCustomer_df['LoyaltyTier'].fillna("Not_Available")

In [39]:
DimCustomer_df.isnull().sum()

CustomerKey             0
CustomerLabel           0
CustomerSegment         0
LoyaltyTier             0
AcquisitionChannel      0
FirstPurchaseDateKey    0
CohortMonth             0
CohortYear              0
CohortQuarter           0
HomeGeographyKey        0
PreferredCategory       0
AgeBand                 0
dtype: int64

In [40]:
# Cheak Column Name Case
DimCustomer_df.columns

Index(['CustomerKey', 'CustomerLabel', 'CustomerSegment', 'LoyaltyTier',
       'AcquisitionChannel', 'FirstPurchaseDateKey', 'CohortMonth',
       'CohortYear', 'CohortQuarter', 'HomeGeographyKey', 'PreferredCategory',
       'AgeBand'],
      dtype='object')

In [41]:
# Convert Column Column Name In Lower Case 
DimCustomer_df.columns = DimCustomer_df.columns.str.lower()

In [42]:
# Cheak Again
DimCustomer_df.columns

Index(['customerkey', 'customerlabel', 'customersegment', 'loyaltytier',
       'acquisitionchannel', 'firstpurchasedatekey', 'cohortmonth',
       'cohortyear', 'cohortquarter', 'homegeographykey', 'preferredcategory',
       'ageband'],
      dtype='object')

### DimCustomer_df Clening is done...!!

### **5. DimGeography_df clening start**

#### cheak head

In [43]:
DimGeography_df.head()

,GeographyKey,Region,Country,City,AOVIndex,ShippingCostIndex,ExpectedOnTimeRate
0,G001,Western Europe,Germany,Berlin,1.00,1.00,0.97
1,G002,Western Europe,Germany,Munich,1.04,1.00,0.97
2,G003,Western Europe,France,Paris,1.06,1.04,0.94
3,G004,Western Europe,France,Lyon,1.00,1.03,0.95
4,G005,Western Europe,Netherlands,Amsterdam,1.05,0.97,0.98


#### cheak info

In [44]:
DimGeography_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24 entries, 0 to 23
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   GeographyKey        24 non-null     object 
 1   Region              24 non-null     object 
 2   Country             24 non-null     object 
 3   City                24 non-null     object 
 4   AOVIndex            24 non-null     float64
 5   ShippingCostIndex   24 non-null     float64
 6   ExpectedOnTimeRate  24 non-null     float64
dtypes: float64(3), object(4)
memory usage: 1.4+ KB


#### cheak null values

In [45]:
DimGeography_df.isna().sum()

GeographyKey          0
Region                0
Country               0
City                  0
AOVIndex              0
ShippingCostIndex     0
ExpectedOnTimeRate    0
dtype: int64

#### cheak statistics

In [46]:
DimGeography_df.describe()

,AOVIndex,ShippingCostIndex,ExpectedOnTimeRate
count,24.000000,24.000000,24.000000
mean,0.973750,1.011250,0.944167
std,0.095089,0.171117,0.036106
min,0.800000,0.760000,0.850000
25%,0.895000,0.837500,0.940000
50%,0.995000,1.015000,0.955000
75%,1.042500,1.140000,0.970000
max,1.130000,1.280000,0.980000


#### cheak duplicate

In [47]:
DimGeography_df.duplicated().sum()

np.int64(0)

#### cheak shape of the data

In [48]:
print("total row =",DimGeography_df.shape[0])
print("total columns =",DimGeography_df.shape[1])

total row = 24
total columns = 7


#### **About the Dataframe**
- No Null values
- No Duplicate values
- Dta Type Is Correct
- Total Row is = 24
- Total Column = 7
- Conver Column Name In Lower Case

In [49]:
# Chek Clumns Name Case
DimGeography_df.columns

Index(['GeographyKey', 'Region', 'Country', 'City', 'AOVIndex',
       'ShippingCostIndex', 'ExpectedOnTimeRate'],
      dtype='object')

In [50]:
# Covert Column Name Into In Lower Case
DimGeography_df.columns = DimGeography_df.columns.str.lower()

In [51]:
# Chek Again 
DimGeography_df.columns

Index(['geographykey', 'region', 'country', 'city', 'aovindex',
       'shippingcostindex', 'expectedontimerate'],
      dtype='object')

### DimGeography_df clening is done....!!

### **6. DimPromotion_df clening start**

#### cheak head

In [52]:
DimPromotion_df.head ()

,PromotionKey,PromotionObjective,Campaign,DiscountType,DefaultDiscountRate,FreeShippingFlag,TrafficQuality
0,PR00,No promotion,No promotion,NaN,0.00,0,Baseline
1,PR01,Retention,Loyalty 10%,Percentage,0.10,0,High
2,PR02,Basket growth,Bundle & Save,Bundle,0.08,0,High
3,PR03,Traffic,Weekend Flash 25%,Percentage,0.25,0,Low
4,PR04,Seasonal demand,Summer Event 15%,Percentage,0.15,0,Medium


#### cheak info

In [53]:
DimPromotion_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 7 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   PromotionKey         10 non-null     object 
 1   PromotionObjective   10 non-null     object 
 2   Campaign             10 non-null     object 
 3   DiscountType         9 non-null      object 
 4   DefaultDiscountRate  10 non-null     float64
 5   FreeShippingFlag     10 non-null     int64  
 6   TrafficQuality       10 non-null     object 
dtypes: float64(1), int64(1), object(5)
memory usage: 692.0+ bytes


#### cheak null values

In [54]:
DimPromotion_df.isna().sum()

PromotionKey           0
PromotionObjective     0
Campaign               0
DiscountType           1
DefaultDiscountRate    0
FreeShippingFlag       0
TrafficQuality         0
dtype: int64

#### cheak statistics

In [55]:
DimPromotion_df.describe()

,DefaultDiscountRate,FreeShippingFlag
count,10.000000,10.000000
mean,0.140000,0.200000
std,0.107909,0.421637
min,0.000000,0.000000
25%,0.085000,0.000000
50%,0.135000,0.000000
75%,0.187500,0.000000
max,0.350000,1.000000


#### cheak duplicate

In [56]:
DimPromotion_df.duplicated().sum()

np.int64(0)

#### cheak shape of the data

In [57]:
print("total row =",DimPromotion_df.shape[0])
print("total columns =",DimPromotion_df.shape[1])

total row = 10
total columns = 7


#### **About the Dataframe**
- Null values - **DiscountType :- 1**
- No Duplicate values
- Dta Type Is Correct
- Total Row is = 10
- Total Column = 7
- Conver Column Name In Lower Case
- fill null values discountype = **No**

In [58]:
DimPromotion_df['DiscountType'] = DimPromotion_df['DiscountType'].fillna("no") 

In [59]:
# Cheak Column name case
DimPromotion_df.columns

Index(['PromotionKey', 'PromotionObjective', 'Campaign', 'DiscountType',
       'DefaultDiscountRate', 'FreeShippingFlag', 'TrafficQuality'],
      dtype='object')

In [60]:
# Convert column name into lower case
DimPromotion_df.columns = DimPromotion_df.columns.str.lower()

In [61]:
# chek again 
DimPromotion_df.columns

Index(['promotionkey', 'promotionobjective', 'campaign', 'discounttype',
       'defaultdiscountrate', 'freeshippingflag', 'trafficquality'],
      dtype='object')

### DimPromotion_df clening is done

### **7. DimFulfillment_df clening start**

#### cheak head

In [62]:
DimFulfillment_df.head()

,FulfillmentKey,FulfillmentModel,FulfillmentCenter,Carrier,ServiceLevel,BasePromisedDays,CostIndex
0,F01,Owned network,Berlin FC,DHL,Standard,3,1.00
1,F02,Owned network,Warsaw FC,DPD,Standard,3,0.93
2,F03,Owned network,Riga FC,Omniva,Economy,3,0.78
3,F04,3PL,Madrid Hub,GLS,Standard,4,1.14
4,F05,3PL,Amsterdam Hub,DHL,Express,2,1.42


In [114]:
DimFulfillment_df.columns.tolist()

['fulfillmentkey',
 'fulfillmentmodel',
 'fulfillmentcenter',
 'carrier',
 'servicelevel',
 'basepromiseddays',
 'costindex']

#### cheak info

In [63]:
DimFulfillment_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   FulfillmentKey     6 non-null      object 
 1   FulfillmentModel   6 non-null      object 
 2   FulfillmentCenter  6 non-null      object 
 3   Carrier            6 non-null      object 
 4   ServiceLevel       6 non-null      object 
 5   BasePromisedDays   6 non-null      int64  
 6   CostIndex          6 non-null      float64
dtypes: float64(1), int64(1), object(5)
memory usage: 468.0+ bytes


#### cheak null values

In [64]:
DimFulfillment_df.isna().sum()

FulfillmentKey       0
FulfillmentModel     0
FulfillmentCenter    0
Carrier              0
ServiceLevel         0
BasePromisedDays     0
CostIndex            0
dtype: int64

#### cheak statistics

In [65]:
DimFulfillment_df.describe()

,BasePromisedDays,CostIndex
count,6.000000,6.000000
mean,3.333333,1.081667
std,1.032796,0.227017
min,2.000000,0.780000
25%,3.000000,0.947500
50%,3.000000,1.070000
75%,3.750000,1.200000
max,5.000000,1.420000


#### cheak duplicate

In [66]:
DimFulfillment_df.duplicated().sum()

np.int64(0)

#### cheak shape of the data

In [67]:
print("total row =",DimFulfillment_df.shape[0])
print("total columns =",DimFulfillment_df.shape[1])

total row = 6
total columns = 7


#### **About the Dataframe**
- No Null values
- No Duplicate values
- Dta Type Is Correct
- Total Row is = 6
- Total Column = 7
- Conver Column Name In Lower Case

In [68]:
# Cheak Column name case
DimFulfillment_df.columns

Index(['FulfillmentKey', 'FulfillmentModel', 'FulfillmentCenter', 'Carrier',
       'ServiceLevel', 'BasePromisedDays', 'CostIndex'],
      dtype='object')

In [69]:
# Convert column name into lower case
DimFulfillment_df.columns = DimFulfillment_df.columns.str.lower()

In [70]:
# chek again 
DimFulfillment_df.columns

Index(['fulfillmentkey', 'fulfillmentmodel', 'fulfillmentcenter', 'carrier',
       'servicelevel', 'basepromiseddays', 'costindex'],
      dtype='object')

### DimFulfillment_df Clening is done

### **8. DimSalesChannel_df clening start**

#### cheak head

In [71]:
DimSalesChannel_df.head() 

,SalesChannelKey,ChannelGroup,SalesChannel,DeviceType,PaymentFeeRate
0,SC01,Owned commerce,Website,Desktop,0.024
1,SC02,Owned commerce,Website,Mobile,0.025
2,SC03,Owned commerce,Mobile App,App,0.021
3,SC04,Marketplace,Marketplace,Mixed,0.082
4,SC05,Social commerce,Social Shop,Mobile,0.055


#### cheak info

In [72]:
DimSalesChannel_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   SalesChannelKey  5 non-null      object 
 1   ChannelGroup     5 non-null      object 
 2   SalesChannel     5 non-null      object 
 3   DeviceType       5 non-null      object 
 4   PaymentFeeRate   5 non-null      float64
dtypes: float64(1), object(4)
memory usage: 332.0+ bytes


#### cheak null values

In [73]:
DimSalesChannel_df.isna().sum()

SalesChannelKey    0
ChannelGroup       0
SalesChannel       0
DeviceType         0
PaymentFeeRate     0
dtype: int64

#### cheak statistics

In [74]:
DimSalesChannel_df.describe()

,PaymentFeeRate
count,5.000000
mean,0.041400
std,0.026557
min,0.021000
25%,0.024000
50%,0.025000
75%,0.055000
max,0.082000


#### cheak duplicate

In [75]:
DimSalesChannel_df.duplicated().sum()

np.int64(0)

#### cheak shape of the data

In [76]:
print("total row =",DimSalesChannel_df.shape[0])
print("total columns =",DimSalesChannel_df.shape[1])

total row = 5
total columns = 5


#### **About the Dataframe**
- No Null values
- No Duplicate values
- Dta Type Is Correct
- Total Row is = 5
- Total Column = 5
- Conver Column Name In Lower Case

In [77]:
# Cheak Column name case
DimSalesChannel_df.columns

Index(['SalesChannelKey', 'ChannelGroup', 'SalesChannel', 'DeviceType',
       'PaymentFeeRate'],
      dtype='object')

In [78]:
# Convert column name into lower case
DimSalesChannel_df.columns = DimSalesChannel_df.columns.str.lower()

In [79]:
# chek again 
DimSalesChannel_df.columns

Index(['saleschannelkey', 'channelgroup', 'saleschannel', 'devicetype',
       'paymentfeerate'],
      dtype='object')

### DimSalesChannel_df clening is done

### **9. DimReturnReason_df clening start**

#### cheak head

In [80]:
DimReturnReason_df.head() 

,ReturnReasonKey,ReturnReasonGroup,ReturnReason,InventoryRecoveryRate
0,R00,No return,Not returned,1.00
1,R01,Fit / preference,Size or fit issue,0.88
2,R02,Fit / preference,Changed mind,0.78
3,R03,Product quality,Damaged in transit,0.10
4,R04,Product quality,Defective product,0.00


#### cheak info

In [81]:
DimReturnReason_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 4 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   ReturnReasonKey        7 non-null      object 
 1   ReturnReasonGroup      7 non-null      object 
 2   ReturnReason           7 non-null      object 
 3   InventoryRecoveryRate  7 non-null      float64
dtypes: float64(1), object(3)
memory usage: 356.0+ bytes


#### cheak null values

In [82]:
DimReturnReason_df.isna().sum()

ReturnReasonKey          0
ReturnReasonGroup        0
ReturnReason             0
InventoryRecoveryRate    0
dtype: int64

#### cheak statistics

In [83]:
DimReturnReason_df.describe()

,InventoryRecoveryRate
count,7.000000
mean,0.575714
std,0.385913
min,0.000000
25%,0.325000
50%,0.720000
75%,0.830000
max,1.000000


#### cheak duplicate

In [84]:
DimReturnReason_df.duplicated().sum()

np.int64(0)

#### cheak shape of the data

In [85]:
print("total row =",DimReturnReason_df.shape[0])
print("total columns =",DimReturnReason_df.shape[1])

total row = 7
total columns = 4


#### **About the Dataframe**
- No Null values
- No Duplicate values
- Dta Type Is Correct
- Total Row is = 5
- Total Column = 5
- Conver Column Name In Lower Case

In [86]:
# Cheak Column name case
DimReturnReason_df.columns

Index(['ReturnReasonKey', 'ReturnReasonGroup', 'ReturnReason',
       'InventoryRecoveryRate'],
      dtype='object')

In [87]:
# Convert column name into lower case
DimReturnReason_df.columns = DimReturnReason_df.columns.str.lower()

In [88]:
# chek again 
DimReturnReason_df.columns

Index(['returnreasonkey', 'returnreasongroup', 'returnreason',
       'inventoryrecoveryrate'],
      dtype='object')

### DimReturnReason_df clening is done

### **10. DimCohortAge_df clening start** 

#### cheak head

In [89]:
DimCohortAge_df.head()

,CohortAgeKey,MonthsSinceFirstPurchase,CohortAgeLabel,LifecycleStage
0,0,0,Month 0,Acquisition
1,1,1,Month 1,Early repeat
2,2,2,Month 2,Early repeat
3,3,3,Month 3,Early repeat
4,4,4,Month 4,Developing


#### cheak info

In [90]:
DimCohortAge_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25 entries, 0 to 24
Data columns (total 4 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   CohortAgeKey              25 non-null     int64 
 1   MonthsSinceFirstPurchase  25 non-null     int64 
 2   CohortAgeLabel            25 non-null     object
 3   LifecycleStage            25 non-null     object
dtypes: int64(2), object(2)
memory usage: 932.0+ bytes


#### cheak null values

In [91]:
DimCohortAge_df.isna().sum()

CohortAgeKey                0
MonthsSinceFirstPurchase    0
CohortAgeLabel              0
LifecycleStage              0
dtype: int64

#### cheak statistics

In [92]:
DimCohortAge_df.describe()

,CohortAgeKey,MonthsSinceFirstPurchase
count,25.000000,25.000000
mean,12.000000,12.000000
std,7.359801,7.359801
min,0.000000,0.000000
25%,6.000000,6.000000
50%,12.000000,12.000000
75%,18.000000,18.000000
max,24.000000,24.000000


#### cheak duplicate

In [93]:
DimCohortAge_df.duplicated().sum()

np.int64(0)

#### cheak shape of the data

In [94]:
print("total row =",DimCohortAge_df.shape[0])
print("total columns =",DimCohortAge_df.shape[1])

total row = 25
total columns = 4


#### **About the Dataframe**
- No Null values
- No Duplicate values
- Dta Type Is Correct
- Total Row is = 25
- Total Column = 4
- Conver Column Name In Lower Case

In [95]:
# Cheak Column name case
DimCohortAge_df.columns

Index(['CohortAgeKey', 'MonthsSinceFirstPurchase', 'CohortAgeLabel',
       'LifecycleStage'],
      dtype='object')

In [96]:
# Convert column name into lower case
DimCohortAge_df.columns = DimCohortAge_df.columns.str.lower()

In [97]:
# Cheak Column name case
DimCohortAge_df.columns

Index(['cohortagekey', 'monthssincefirstpurchase', 'cohortagelabel',
       'lifecyclestage'],
      dtype='object')

### DimCohortAge_df clening is done 

### **11. DataDictionary_df clening start**

#### cheak head

In [98]:
DataDictionary_df.head()

,Table,Column,Data type,Default aggregation,Definition / modeling note
0,FactOrderLine,LineID,Text,No,Unique requested order-line identifier.
1,FactOrderLine,OrderID,Text,Distinct count,Order identifier; multiple lines can share one...
2,FactOrderLine,OrderDateKey,Whole number,No,Active date foreign key.
3,FactOrderLine,ShipDateKey,Whole number,No,Inactive role-playing date key.
4,FactOrderLine,DeliveryDateKey,Whole number,No,Inactive role-playing date key; blank for unfu...


### **this table is all about informations**

## All Clening Procces Is Done

### **Save Data Into In Csv File**

## **Load All Data In PosrgreSql using (sqlalchemy)**

DimDate_df.to_sql("dimdate", engine, if_exists="append", index=False)
print("dimdate loaded successfully!")

# **All Is Done**
- **lets continute in sql for better analysis**

# **Prakrit singh analyst**